xds/csds/csds_e2e_test.go

```diff
@@ -65,56 +64,113 @@ func Test(t *testing.T) {
 	grpctest.RunSubTests(t, s{})
 }
 
-// The following watcher implementations are no-ops since we don't really care
-// about the callback received by these watchers in the test. We only care
-// whether CSDS reports the expected state.
+// The following watcher implementations are no-ops in the sense that they don't
+// validate or use the received updates from the xDS client. In this test, we
+// only care whether CSDS reports the expected state.
+//
+// Note: These watchers do use the ADS stream level flow control supported by
+// the xDS client. The watch callbacks block on a wait channel that is written
+// to by the test, before invoking the onDone callback. This is particularly
+// useful when a resource is NACKed, because the go-control-plane management
+// server continuously resends the same resource in this case, and applying flow
+// control from these watchers ensures that xDS client does not spend all of its
+// time receiving and NACKing updates from the maangement server. This was
+// indeed the case on arm64 (before we had support for ADS stream level flow
+// control), and was causing CSDS to not receive any updates in this case.
+
+func blockAndDone(testDoneCh <-chan struct{}, waitCh chan struct{}, onDone xdsresource.DoneNotifier) {
+	select {
+	case <-testDoneCh:
+	case <-waitCh:
+		onDone.OnDone()
+	}
+}
+
+type blockingListenerWatcher struct {
+	testDoneCh <-chan struct{} // Closed when the test is done.
+	waitCh     chan struct{}   // Written to by the test to unblock the watch callback.
```

Comment:
Nit: here, other watchers, and the helper unblockResourceWatchers. "wait" to me is encoded implicitly in the semantics of a go channel. Perhaps call this onDoneCh? triggerOnDoneCh? watchCallbackCh?

Reply:
Done. Thanks.